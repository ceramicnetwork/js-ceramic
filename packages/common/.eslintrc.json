{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint", "import"],
  "rules": {
    "@typescript-eslint/no-explicit-any": "off",
    "@typescript-eslint/ban-ts-ignore": "off",
    "@typescript-eslint/ban-ts-comment": "off",
    "import/no-unresolved": "error",
    "import/default": "off"
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [".ts", ".tsx"]
    },
    "import/resolver": {
      "typescript": {
        "extensions": []
      },
      "node": true
    }
  },
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:import/typescript"
  ]
}
