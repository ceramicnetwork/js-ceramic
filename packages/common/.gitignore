node_modules/
coverage/
lib/
src/generated/
