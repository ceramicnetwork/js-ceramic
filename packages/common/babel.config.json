{
  "presets": [
    [
      "@babel/preset-env",
      {
        "targets": {
          "node": "16"
        },
        "useBuiltIns": false,
        "debug": false
      }
    ],
    "@babel/preset-typescript"
  ],
  "plugins": [
    [
      "@babel/plugin-transform-modules-commonjs",
      {
        "allowTopLevelThis": true
      }
    ],
    ["@babel/plugin-proposal-decorators", { "legacy": true }]
  ]
}
