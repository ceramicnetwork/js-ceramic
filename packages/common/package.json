{
  "name": "@ceramicnetwork/common",
  "version": "0.13.0",
  "description": "Ceramic common types and utilities",
  "keywords": [
    "ceramic",
    "types",
    "utilities"
  ],
  "author": "Janko Simonovic <simonovic86@gmail.com>",
  "homepage": "",
  "license": "(Apache-2.0 OR MIT)",
  "directories": {
    "lib": "lib"
  },
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib"
  ],
  "scripts": {
    "test": "./node_modules/.bin/jest --coverage --passWithNoTests",
    "build": "../../node_modules/.bin/tsc -p tsconfig.json",
    "prepublishOnly": "npm run build",
    "prebuild": "npm run clean",
    "lint": "./node_modules/.bin/eslint ./src --ext .js,.jsx,.ts,.tsx",
    "clean": "rm -rf ./lib"
  },
  "dependencies": {
    "@ceramicnetwork/docid": "^0.3.0",
    "cids": "^1.0.2",
    "dag-jose": "^0.3.0",
    "did-resolver": "^2.1.1",
    "lodash.clonedeep": "^4.5.0",
    "loglevel": "^1.7.0",
    "loglevel-plugin-prefix": "^0.8.4",
    "uint8arrays": "^1.1.0"
  },
  "devDependencies": {
    "@babel/core": "^7.9.0",
    "@babel/plugin-proposal-decorators": "^7.10.1",
    "@babel/preset-env": "^7.8.4",
    "@babel/preset-typescript": "^7.9.0",
    "@types/encoding-down": "^5.0.0",
    "@types/events": "^3.0.0",
    "@types/json-schema": "^7.0.5",
    "@types/lodash.clonedeep": "^4.5.6",
    "@types/node": "^13.13.15",
    "@types/prettier": "^2.0.2",
    "@typescript-eslint/eslint-plugin": "^4.6.0",
    "@typescript-eslint/parser": "^4.6.0",
    "babel-jest": "^25.1.0",
    "dids": "^1.0.0",
    "eslint": "^6.8.0",
    "eslint-plugin-jest": "^23.8.2",
    "jest": "^25.1.0",
    "js-sha256": "^0.9.0",
    "json-schema-to-typescript": "^9.1.1",
    "tmp-promise": "^2.0.2",
    "ts-node": "^9.0.0",
    "tsc-watch": "^4.2.9",
    "ipfs-core": "^0.3.0",
    "typescript-json-schema": "^0.42.0"
  },
  "jest": {
    "transformIgnorePatterns": [
      "/node_modules(?!/did-jwt)/"
    ]
  },
  "gitHead": "c58eb9659f5fdd976aca2db6465619c20ba30ff9"
}
