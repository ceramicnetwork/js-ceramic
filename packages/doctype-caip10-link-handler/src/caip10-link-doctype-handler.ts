import type CID from 'cids'
import { validateLink } from "@ceramicnetwork/blockchain-utils-validation"
import { Caip10LinkDoctype, Caip10LinkParams, DOCTYPE_NAME } from "@ceramicnetwork/doctype-caip10-link"
import {
    AnchorStatus,
    DocState,
    DoctypeConstructor,
    DoctypeHandler,
    DocOpts,
    SignatureStatus,
    CommitType,
    CeramicCommit,
    AnchorCommit,
    Context
} from "@ceramicnetwork/common"

const IPFS_GET_TIMEOUT = 60000 // 1 minute

export class Caip10LinkDoctypeHandler implements DoctypeHandler<Caip10LinkDoctype> {
    /**
     * Gets doctype name
     */
    get name(): string {
        return DOCTYPE_NAME
    }

    /**
     * Gets doctype class
     */
    get doctype(): DoctypeConstructor<Caip10LinkDoctype> {
        return Caip10LinkDoctype
    }

    /**
     * Creates Caip10Link instance
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    async create(params: Caip10LinkParams, context: Context, opts?: DocOpts): Promise<Caip10LinkDoctype> {
        return Caip10LinkDoctype.create(params, context, opts);
    }

    /**
     * Applies commit (genesis|signed|anchor)
     * @param commit - Commit to be applied
     * @param cid - Commit CID
     * @param context - Ceramic context
     * @param state - Document state
     */
    async applyCommit(commit: CeramicCommit, cid: CID, context: Context, state?: DocState): Promise<DocState> {
        if (state == null) {
            return this._applyGenesis(commit, cid)
        }

        if ((commit as AnchorCommit).proof) {
            return this._applyAnchor(context, commit, cid, state);
        }

        return this._applySigned(commit, cid, state);
    }

    /**
     * Applies genesis commit
     * @param commit - Genesis commit
     * @param cid - Genesis commit CID
     * @private
     */
    async _applyGenesis (commit: any, cid: CID): Promise<DocState> {
        // TODO - verify genesis commit
        const state = {
            doctype: DOCTYPE_NAME,
            content: null,
            next: {
                content: null
            },
            metadata: commit.header,
            signature: SignatureStatus.GENESIS,
            anchorStatus: AnchorStatus.NOT_REQUESTED,
            log: [{ cid, type: CommitType.GENESIS }]
        }

        if (!(state.metadata.controllers && state.metadata.controllers.length === 1)) {
            throw new Error('Exactly one controller must be specified')
        }

        return state
    }

    /**
     * Applies signed commit
     * @param commit - Signed commit
     * @param cid - Signed commit CID
     * @param state - Document state
     * @private
     */
    async _applySigned (commit: any, cid: CID, state: DocState): Promise<DocState> {
        // TODO: Assert that the 'prev' of the commit being applied is the end of the log in 'state'
        const validProof = await validateLink(commit.data)
        if (!validProof) {
            throw new Error('Invalid proof for signed commit')
        }

        if (state.signature !== SignatureStatus.GENESIS && (
          (
            state.anchorStatus === AnchorStatus.ANCHORED &&
            validProof.timestamp < state.anchorProof.blockTimestamp
          ) || (
            state.anchorStatus !== AnchorStatus.ANCHORED &&
            validProof.timestamp < state.next.metadata.lastUpdate
          )
        )) {
          throw new Error('Invalid commit, proof timestamp too old')
        }

        // TODO: handle CAIP-10 addresses in proof generation of 3id-blockchain-utils
        const account = validProof.account || validProof.address
        let [address, chainId] = account.split('@')  // eslint-disable-line prefer-const

        const addressCaip10 = [address, chainId].join('@')
        if (addressCaip10.toLowerCase() !== state.metadata.controllers[0].toLowerCase()) {
            throw new Error("Address doesn't match document controller")
        }
        state.log.push({ cid, type: CommitType.SIGNED })
        return {
            ...state,
            signature: SignatureStatus.SIGNED,
            anchorStatus: AnchorStatus.NOT_REQUESTED,
            next: {
                content: validProof.did,
                metadata: {
                  ...state.metadata,
                  lastUpdate: validProof.timestamp // in case there are two updates after each other
                }
            }
        }
    }

    /**
     * Applies anchor commit
     * @param context - Ceramic context
     * @param commit - Anchor commit
     * @param cid - Anchor commit CID
     * @param state - Document state
     * @private
     */
    async _applyAnchor (context: Context, commit: any, cid: CID, state: DocState): Promise<DocState> {
        // TODO: Assert that the 'prev' of the commit being applied is the end of the log in 'state'
        const proof = (await context.ipfs.dag.get(commit.proof, { timeout: IPFS_GET_TIMEOUT })).value;

        const supportedChains = await context.api.getSupportedChains()
        if (!supportedChains.includes(proof.chainId)) {
            throw new Error("Anchor proof chainId '" + proof.chainId
                + "' is not supported. Supported chains are: '"
                + supportedChains.join("', '") + "'")
        }

        state.log.push({ cid, type: CommitType.ANCHOR })
        let content = state.content
        if (state.next?.content) {
            content = state.next.content
            delete state.next.content
            delete state.next.metadata
        }
        return {
            ...state,
            content,
            anchorStatus: AnchorStatus.ANCHORED,
            anchorProof: proof,
        }
    }

}
