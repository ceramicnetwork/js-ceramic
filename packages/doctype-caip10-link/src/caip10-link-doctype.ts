import { Doctype, DoctypeConstructor, DoctypeStatic, DocOpts, DocParams, Context } from "@ceramicnetwork/common"

const DOCTYPE = 'caip10-link'

/**
 * Caip10Link parameters
 */
export interface Caip10LinkParams extends DocParams {
    content?: Record<string, unknown>;
}

/**
 * Caip10Link doctype implementation
 */
@DoctypeStatic<DoctypeConstructor<Caip10LinkDoctype>>()
export class Caip10LinkDoctype extends Doctype {

    /**
     * Changes Caip10Link instance
     * @param params - Change parameters
     * @param opts - Initialization options
     */
    async change(params: Caip10LinkParams, opts?: DocOpts): Promise<void> {
        const { content, metadata } = params
        const updateRecord = await Caip10LinkDoctype._makeRecord(this, content, metadata?.schema)
        const updated = await this.context.api.applyRecord(this.id.toString(), updateRecord, opts)
        this.state = updated.state
    }

    /**
     * Creates Caip10Link doctype
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    static async create(params: Caip10LinkParams, context: Context, opts?: DocOpts): Promise<Caip10LinkDoctype> {
        const { content, metadata } = params

        const record = await Caip10LinkDoctype.makeGenesis({ content, metadata })
        return context.api.createDocumentFromGenesis(DOCTYPE, record, opts)
    }

    /**
     * Creates genesis record
     * @param params - Create parameters
     */
    static async makeGenesis(params: Record<string, any>): Promise<Record<string, any>> {
        const { content, metadata } = params

        if (content) {
            throw new Error('Account link genesis cannot have content')
        }
        if (!metadata) {
            throw new Error('Metadata must be specified')
        }
        if (!metadata.controllers) {
            throw new Error('Controller must be specified')
        }
        if (metadata.controllers.length !== 1) {
            throw new Error('Exactly one controller must be specified')
        }
        const [address, chainId] = metadata.controllers[0].split('@') // eslint-disable-line @typescript-eslint/no-unused-vars
        if (!chainId) {
            throw new Error('Chain ID must be specified according to CAIP-10')
        }
        return { header: metadata }
    }

    /**
     * Creates change record
     * @param doctype - Caip10Link doctype instance
     * @param newContent - Change content
     * @param newSchema - Change schema
     * @private
     */
    static async _makeRecord (doctype: Caip10LinkDoctype, newContent: any, newSchema: string = null): Promise<any> {
        const { metadata } = doctype
        if (newSchema) {
            metadata.schema = newSchema
        }
        if (newContent == null) {
            newContent = doctype.content
        }
        return { data: newContent, header: metadata, prev: doctype.tip, id: doctype.state.log[0].cid }
    }
}
