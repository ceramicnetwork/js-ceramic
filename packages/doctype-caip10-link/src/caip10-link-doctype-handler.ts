import CID from 'cids'
import { validateLink } from "3id-blockchain-utils"
import { Caip10LinkDoctype, Caip10LinkParams } from "./caip10-link-doctype"
import {
    AnchorStatus,
    DocState,
    DoctypeConstructor,
    DoctypeHandler,
    DocOpts,
    SignatureStatus,
    RecordType,
    CeramicRecord,
    AnchorRecord,
} from "@ceramicnetwork/common"
import { Context } from "@ceramicnetwork/common"

const DOCTYPE = 'caip10-link'

export class Caip10LinkDoctypeHandler implements DoctypeHandler<Caip10LinkDoctype> {
    /**
     * Gets doctype name
     */
    get name(): string {
        return DOCTYPE
    }

    /**
     * Gets doctype class
     */
    get doctype(): DoctypeConstructor<Caip10LinkDoctype> {
        return Caip10LinkDoctype
    }

    /**
     * Creates Caip10Link instance
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    async create(params: Caip10LinkParams, context: Context, opts?: DocOpts): Promise<Caip10LinkDoctype> {
        return Caip10LinkDoctype.create(params, context, opts);
    }

    /**
     * Applies record (genesis|signed|anchor)
     * @param record - Record to be applied
     * @param cid - Record CID
     * @param context - Ceramic context
     * @param state - Document state
     */
    async applyRecord(record: CeramicRecord, cid: CID, context: Context, state?: DocState): Promise<DocState> {
        if (state == null) {
            return this._applyGenesis(record, cid)
        }

        if ((record as AnchorRecord).proof) {
            return this._applyAnchor(context, record, cid, state);
        }

        return this._applySigned(record, cid, state);
    }

    /**
     * Applies genesis record
     * @param record - Genesis record
     * @param cid - Genesis record CID
     * @private
     */
    async _applyGenesis (record: any, cid: CID): Promise<DocState> {
        // TODO - verify genesis record
        return {
            doctype: DOCTYPE,
            content: null,
            next: {
                content: null
            },
            metadata: record.header,
            signature: SignatureStatus.GENESIS,
            anchorStatus: AnchorStatus.NOT_REQUESTED,
            log: [{ cid, type: RecordType.GENESIS }]
        }
    }

    /**
     * Applies signed record
     * @param record - Signed record
     * @param cid - Signed record CID
     * @param state - Document state
     * @private
     */
    async _applySigned (record: any, cid: CID, state: DocState): Promise<DocState> {
        // TODO: Assert that the 'prev' of the record being applied is the end of the log in 'state'
        const validProof = await validateLink(record.data)
        if (!validProof) {
            throw new Error('Invalid proof for signed record')
        }

        // TODO: handle CAIP-10 addresses in proof generation of 3id-blockchain-utils
        const account = validProof.account || validProof.address
        let [address, chainId] = account.split('@')  // eslint-disable-line prefer-const

        const addressCaip10 = [address, chainId].join('@')
        if (addressCaip10.toLowerCase() !== state.metadata.controllers[0].toLowerCase()) {
            throw new Error("Address doesn't match document controller")
        }
        state.log.push({ cid, type: RecordType.SIGNED })
        return {
            ...state,
            signature: SignatureStatus.SIGNED,
            anchorStatus: AnchorStatus.NOT_REQUESTED,
            next: {
                content: validProof.did
            }
        }
    }

    /**
     * Applies anchor record
     * @param context - Ceramic context
     * @param record - Anchor record
     * @param cid - Anchor record CID
     * @param state - Document state
     * @private
     */
    async _applyAnchor (context: Context, record: any, cid: CID, state: DocState): Promise<DocState> {
        // TODO: Assert that the 'prev' of the record being applied is the end of the log in 'state'
        const proof = (await context.ipfs.dag.get(record.proof)).value;

        const supportedChains = await context.api.getSupportedChains()
        if (!supportedChains.includes(proof.chainId)) {
            throw new Error("Anchor proof chainId '" + proof.chainId
                + "' is not supported. Supported chains are: '"
                + supportedChains.join("', '") + "'")
        }

        state.log.push({ cid, type: RecordType.ANCHOR })
        let content = state.content
        if (state.next?.content) {
            content = state.next.content
            delete state.next.content
        }
        return {
            ...state,
            content,
            anchorStatus: AnchorStatus.ANCHORED,
            anchorProof: proof,
        }
    }

}
