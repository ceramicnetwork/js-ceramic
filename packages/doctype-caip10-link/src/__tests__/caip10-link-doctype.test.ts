import { Caip10LinkDoctypeHandler } from '../caip10-link-doctype-handler'

import cloneDeep from 'lodash.clonedeep'
import CID from 'cids'

import * as u8a from 'uint8arrays'
import { validateLink, createLink } from '3id-blockchain-utils'
import { Caip10LinkDoctype } from "../caip10-link-doctype"
import { Context } from "@ceramicnetwork/common"

const { sha256 } = require('js-sha256') // eslint-disable-line @typescript-eslint/no-var-requires
const hash = (data: string): CID => new CID(1, 'sha2-256', Buffer.from('1220' + sha256(data), 'hex'))

const FAKE_CID_1 = new CID('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_2 = new CID('bafybeig6xv5nwphfmvcnektpnojts44jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_3 = new CID('bafybeig6xv5nwphfmvcnektpnojts55jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_4 = new CID('bafybeig6xv5nwphfmvcnektpnojts66jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_5 = new CID('bafybeig6xv5nwphfmvcnektpnojts44jqcuam7bmye2pb54adnrtccjlse')
const FAKE_CID_6 = new CID('bafybeig6xv5nwphfmvcnektpnojts55jqcuam7bmye2pb54adnrtccjleu')
const FAKE_CID_7 = new CID('bafybeig6xv5nwphfmvcnektpnojts66jqcuam7bmye2pb54adnrtccjlwu')
const FAKE_CID_8 = new CID('bafybeig6xv5nwphfmvcnektpnojts66jqcuam6bmye2pb54adnrtccjlwu')

const RECORDS = {
  genesis: { header: { controllers: [ '0x0544DcF4fcE959C6C4F3b7530190cB5E1BD67Cb8@eip155:1' ], family: "caip10-eip155:1" } },
  r1: {
    desiredContent: {
      version: 2,
      type: 'ethereum-eoa',
      message: 'Create a new account link to your identity.\n' + '\n' + 'did:3:testdid1 \n' + 'Timestamp: 1608116721',
      signature: '0xc6a5f50945bc7b06320b66cfe144e2b571391c88827eed0490f7f8e5e8af769c4246e27e83078c907ccfaf9bd0de82bd5e57fe7884d9cb8a9303f5d92ea4df0d1c',
      account: '0x0544dcf4fce959c6c4f3b7530190cb5e1bd67cb8@eip155:1',
      timestamp: 1608116721
    },
    record: {
      data: {
        version: 2,
        type: 'ethereum-eoa',
        message: 'Create a new account link to your identity.\n' + '\n' + 'did:3:testdid1 \n' + 'Timestamp: 1608116721',
        signature: '0xc6a5f50945bc7b06320b66cfe144e2b571391c88827eed0490f7f8e5e8af769c4246e27e83078c907ccfaf9bd0de82bd5e57fe7884d9cb8a9303f5d92ea4df0d1c',
        account: '0x0544dcf4fce959c6c4f3b7530190cb5e1bd67cb8@eip155:1',
        timestamp: 1608116721
      },
      id: FAKE_CID_1,
      prev: FAKE_CID_1
    }
  },
  r2: { record: { proof: FAKE_CID_4 } },
  proof: {
    value: {
      blockNumber: 123456,
      chainId: 'fakechain:123',
    }
  }
}

describe('Caip10LinkHandler', () => {
  let context: Context
  let handler: Caip10LinkDoctypeHandler

  beforeEach(() => {
    handler = new Caip10LinkDoctypeHandler()
    const recs: Record<string, any> = {}
    const ipfs = {
      dag: {
        put(rec: any, cid?: CID): any {
          if (cid) {
            recs[cid.toString()] = rec
            return cid
          }
          // stringify as a way of doing deep copy
          const clone = cloneDeep(rec)
          const c = hash(JSON.stringify(clone))
          recs[c.toString()] = clone
          return c
        },
        get(cid: any): any {
          return recs[cid.toString()]
        }
      }
    }

    const api = {getSupportedChains: jest.fn(async () => {return ["fakechain:123"]})}
    context = {
      ipfs: ipfs,
      anchorService: null,
      api,
    }
  })

  it('is constructed correctly', async () => {
    expect(handler.name).toEqual('caip10-link')
  })

  it('makes genesis record correctly', async () => {
    const record = await Caip10LinkDoctype.makeGenesis({ content: undefined, metadata: RECORDS.genesis.header }, context)
    expect(record).toEqual(RECORDS.genesis)
  })

  it('throws an error if genesis record has content', async () => {
    const content = {}
    await expect(Caip10LinkDoctype.makeGenesis({ content }, context)).rejects.toThrow(/Cannot have content/i)
  })

  it('throws an error if genesis record has no metadata specified', async () => {
    const content: any = undefined
    const controllers: any = undefined
    await expect(Caip10LinkDoctype.makeGenesis({ content, controllers }, context)).rejects.toThrow(/Metadata must be specified/i)
  })

  it('throws an error if genesis record has no controllers specified', async () => {
    const content: any = undefined
    await expect(Caip10LinkDoctype.makeGenesis({ content, metadata: {} }, context)).rejects.toThrow(/Controller must be specified/i)
  })

  it('throws an error if genesis record has more than one controller', async () => {
    const content: any = undefined
    const controllers = [...RECORDS.genesis.header.controllers, '0x25954ef14cebbc9af3d79876489a9cfe87043f20@eip155:1']
    await expect(Caip10LinkDoctype.makeGenesis({ content, metadata: { controllers } }, context)).rejects.toThrow(/Exactly one controller/i)
  })

  it('throws an error if genesis record has controller not in CAIP-10 format', async () => {
    const content: any = undefined
    const controllers = RECORDS.genesis.header.controllers.map(address => address.split('@')[0])
    await expect(Caip10LinkDoctype.makeGenesis({ content, metadata: { controllers } }, context)).rejects.toThrow(/According to CAIP-10/i)
  })

  it('applies genesis record correctly', async () => {
    const state = await handler.applyRecord(RECORDS.genesis, FAKE_CID_1, context)
    expect(state).toMatchSnapshot()
  })

  it('makes signed record correctly', async () => {
    const state = await handler.applyRecord(RECORDS.genesis, FAKE_CID_1, context)
    const doctype = new Caip10LinkDoctype(state, context)
    const record = await Caip10LinkDoctype._makeRecord(doctype, RECORDS.r1.desiredContent)
    expect(record).toEqual(RECORDS.r1.record)
  })

  it('applies signed record correctly', async () => {
    let state = await handler.applyRecord(RECORDS.genesis, FAKE_CID_1, context)
    state = await handler.applyRecord(RECORDS.r1.record, FAKE_CID_2, context, state)
    expect(state).toMatchSnapshot()
  })

  it('throws an error of the proof is invalid', async () => {
    const badRecord = cloneDeep(RECORDS.r1.record)
    badRecord.data.signature = '0xc6a5f50945bc7b06320b66cfe144e2b571391c88827eed0490f7f8e5e8af769c4246e27e8302348762387462387648726346877884d9cb8a9303f5d92ea4df0d1c'
    const state = await handler.applyRecord(RECORDS.genesis, FAKE_CID_1, context)
    await expect(handler.applyRecord(badRecord, FAKE_CID_2, context, state)).rejects.toThrow(/Invalid proof/i)
  })

  it('throws an error of the proof doesn\'t match the controller', async () => {
    const badAddressGenesis = cloneDeep(RECORDS.genesis)
    badAddressGenesis.header.controllers = ['0xffffffffffffffffffffffffffffffffffffffff@eip155:1']
    const state = await handler.applyRecord(badAddressGenesis, FAKE_CID_1, context)
    await expect(handler.applyRecord(RECORDS.r1.record, FAKE_CID_2, context, state)).rejects.toThrow(/Address doesn't match/i)
  })

  it('applies anchor record correctly', async () => {
    // create signed record
    await context.ipfs.dag.put(RECORDS.r1.record, FAKE_CID_2)
    // create anchor record
    await context.ipfs.dag.put(RECORDS.r2.record, FAKE_CID_3)
    // create anchor proof
    await context.ipfs.dag.put(RECORDS.proof, FAKE_CID_4)

    // Apply genesis
    let state = await handler.applyRecord(RECORDS.genesis, FAKE_CID_1, context)
    // Apply signed record
    state = await handler.applyRecord(RECORDS.r1.record, FAKE_CID_2, context, state)
    // Apply anchor record
    state = await handler.applyRecord(RECORDS.r2.record, FAKE_CID_3, context, state)
    expect(state).toMatchSnapshot()
  })

  it('Does not apply anchor record on unsupported chain', async () => {
    // create signed record
    await context.ipfs.dag.put(RECORDS.r1.record, FAKE_CID_2)
    // create anchor record
    await context.ipfs.dag.put(RECORDS.r2.record, FAKE_CID_3)
    // create anchor proof with a different chainId than what's in the genesis record
    await context.ipfs.dag.put({value: { blockNumber: 123456, chainId: 'thewrongchain'}}, FAKE_CID_4)

    // Apply genesis
    let state = await handler.applyRecord(RECORDS.genesis, FAKE_CID_1, context)
    // Apply signed record
    state = await handler.applyRecord(RECORDS.r1.record, FAKE_CID_2, context, state)
    // Apply anchor record
    await expect(handler.applyRecord(RECORDS.r2.record, FAKE_CID_3, context, state))
        .rejects.toThrow("Anchor proof chainId 'thewrongchain' is not supported. Supported chains are: 'fakechain:123'")
  })

  it('Should not allow replay attack', async () => {
    const records = {
      genesis: { header: { controllers: [ '0x0544DcF4fcE959C6C4F3b7530190cB5E1BD67Cb8@eip155:1' ], family: "caip10-eip155:1" } },
      r1: {
        data: {
          version: 2,
          type: 'ethereum-eoa',
          message: 'Create a new account link to your identity.\n' + '\n' + 'did:3:testdid1 \n' + 'Timestamp: 1608116721',
          signature: '0xc6a5f50945bc7b06320b66cfe144e2b571391c88827eed0490f7f8e5e8af769c4246e27e83078c907ccfaf9bd0de82bd5e57fe7884d9cb8a9303f5d92ea4df0d1c',
          account: '0x0544dcf4fce959c6c4f3b7530190cb5e1bd67cb8@eip155:1',
          timestamp: 1608116721
        },
        id: FAKE_CID_1,
        prev: FAKE_CID_1
      },
      r2: { proof: FAKE_CID_4 },
      r2proof: {
        value: {
          blockNumber: 123456,
          blockTimestamp: 1608116723,
          chainId: 'fakechain:123',
        }
      },
      r3: {
        data: {
          version: 2,
          type: 'ethereum-eoa',
          message: 'Create a new account link to your identity.\n' + '\n' + 'did:3:testdid2 \n' + 'Timestamp: 1608116725',
          signature: '0xe1bcc3f3b67ae303cb95db9d2d266f95984fd76bf3a4452dbb64ad4d3941998f4cb37f85197c74f9fb5cdf33e9042949a5452a204db2d48d85929406f64aac811b',
          account: '0x0544dcf4fce959c6c4f3b7530190cb5e1bd67cb8@eip155:1',
          timestamp: 1608116725
        },
        id: FAKE_CID_1,
        prev: FAKE_CID_3
      },
      r4: { proof: FAKE_CID_7 },
      r4proof: {
        value: {
          blockNumber: 123456,
          blockTimestamp: 1608116727,
          chainId: 'fakechain:123',
        }
      }
    }
    await context.ipfs.dag.put(records.r2proof, FAKE_CID_4)
    await context.ipfs.dag.put(records.r4proof, FAKE_CID_7)

    let state = await handler.applyRecord(records.genesis, FAKE_CID_1, context)
    state = await handler.applyRecord(records.r1, FAKE_CID_2, context, state)
    state = await handler.applyRecord(records.r2, FAKE_CID_3, context, state)
    expect(state.content).toEqual('did:3:testdid1')
    state = await handler.applyRecord(records.r3, FAKE_CID_5, context, state)

    // create a fake update based on the r1 data to try a replay attack
    const r4 = {
      data: records.r1.data,
      id: FAKE_CID_1,
      prev: FAKE_CID_5
    }
    await expect(handler.applyRecord(r4, FAKE_CID_8, context, state)).rejects.toThrow('Invalid record, proof timestamp too old')

    state = await handler.applyRecord(records.r4, FAKE_CID_6, context, state)
    expect(state.content).toEqual('did:3:testdid2')

    // create a fake update based on the r1 data to try a replay attack
    const r5 = {
      data: records.r1.data,
      id: FAKE_CID_1,
      prev: FAKE_CID_6
    }
    await expect(handler.applyRecord(r5, FAKE_CID_8, context, state)).rejects.toThrow('Invalid record, proof timestamp too old')
  })
})
