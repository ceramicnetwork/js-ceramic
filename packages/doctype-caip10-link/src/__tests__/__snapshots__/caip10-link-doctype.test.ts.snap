// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Caip10LinkHandler applies anchor record correctly 1`] = `
Object {
  "anchorProof": Object {
    "blockNumber": 123456,
    "chainId": "fakechain:123",
  },
  "anchorStatus": 3,
  "content": "did:3:testdid1",
  "doctype": "caip10-link",
  "log": Array [
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          115,
          137,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 1,
    },
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          119,
          169,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 2,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "0x0544DcF4fcE959C6C4F3b7530190cB5E1BD67Cb8@eip155:1",
    ],
    "family": "caip10-eip155:1",
  },
  "next": Object {},
  "signature": 2,
}
`;

exports[`Caip10LinkHandler applies genesis record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": null,
  "doctype": "caip10-link",
  "log": Array [
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "0x0544DcF4fcE959C6C4F3b7530190cB5E1BD67Cb8@eip155:1",
    ],
    "family": "caip10-eip155:1",
  },
  "next": Object {
    "content": null,
  },
  "signature": 0,
}
`;

exports[`Caip10LinkHandler applies signed record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": null,
  "doctype": "caip10-link",
  "log": Array [
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          115,
          137,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 1,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "0x0544DcF4fcE959C6C4F3b7530190cB5E1BD67Cb8@eip155:1",
    ],
    "family": "caip10-eip155:1",
  },
  "next": Object {
    "content": "did:3:testdid1",
    "metadata": Object {
      "controllers": Array [
        "0x0544DcF4fcE959C6C4F3b7530190cB5E1BD67Cb8@eip155:1",
      ],
      "family": "caip10-eip155:1",
      "lastUpdate": 1608116721,
    },
  },
  "signature": 2,
}
`;
